doc/man-pages/man8/read_tape.8
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "READ_TAPE 8"
.TH READ_TAPE 8 "2010-02-11" "OpenAFS" "AFS Command Reference"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
read_tape \- Reads volume dumps from a backup tape to a file
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBread_tape\fR \fB\-tape\fR\ <\fItape\ device\fR>
   \fB\-restore\fR\ <\fI#\ of\ volumes\ to\ restore\fR>
   \fB\-skip\fR\ <\fI#\ of\ volumes\ to\ skip\fR>
   \fB\-file\fR\ <\fIfilename\fR> [\fB\-scan\fR] [\fB\-noask\fR] [\fB\-label\fR]
   [\fB\-vheaders\fR] [\fB\-verbose\fR] [\fB\-help\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBread_tape\fR reads an OpenAFS backup tape and prompts for each dump file to
save.  This command does not require any OpenAFS infrastructure.  This
command does not need an OpenAFS client or server to be available, which is
not the case with the \fIbackup\fR\|(8) command.
.PP
The dump files will be named for the Read/Write name of the volume restored.
After saving each dump file, \fBvos restore\fR or \fBrestorevol\fR can be used to
restore the volume into \s-1AFS\s0 and non-AFS space respectively.
.PP
\&\fBread_tape\fR reads the tape while skipping the specified number of volumes.
After that, it restores the specified number of volumes.  \fBread_tape\fR
doesn't rewind the tape so that it may be used multiple times in succession.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-tape\fR <\fItape device\fR>" 4
.IX Item "-tape <tape device>"
Specifies the tape device from which to restore.
.IP "\fB\-restore\fR <\fI# of volumes to restore\fR>" 4
.IX Item "-restore <# of volumes to restore>"
Specifies the number of volumes to restore from tape.
.IP "\fB\-skip\fR <\fI# of volumes to skip\fR>" 4
.IX Item "-skip <# of volumes to skip>"
Specifies the number of volumes to skip before starting the restore.
.IP "\fB\-file\fR <\fIfilename\fR>" 4
.IX Item "-file <filename>"
Specifies an alternate name for the restored volume dump file rather than
the default of the volume name.
.IP "\fB\-scan\fR" 4
.IX Item "-scan"
Scans the tape.
.IP "\fB\-noask\fR" 4
.IX Item "-noask"
Doesn't prompt for each volume.
.IP "\fB\-label\fR" 4
.IX Item "-label"
Displays the full dump label.
.IP "\fB\-vheaders\fR" 4
.IX Item "-vheaders"
Displays the full volume headers.
.IP "\fB\-verbose\fR" 4
.IX Item "-verbose"
Produces on the standard output stream a detailed trace of the command's
execution.  If this argument is omitted, only warnings and error messages
appear.
.IP "\fB\-help\fR" 4
.IX Item "-help"
Prints the online help for this command. All other valid options are
ignored.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
The following command will read the third through fifth volumes from
the tape device /dev/tape without prompting:
.PP
.Vb 1
\&   % read_tape \-tape /dev/tape \-skip 2 \-restore 3 \-noask
.Ve
.SH "PRIVILEGE REQUIRED"
.IX Header "PRIVILEGE REQUIRED"
The issuer must have access to read and write to the specified tape device.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIbackup\fR\|(8),
\&\fIrestorevol\fR\|(1),
\&\fIvos_restore\fR\|(1)
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright 2007 Jason Edgecombe <jason@rampaginggeek.com>
.PP
This documentation is covered by the \s-1BSD\s0 License as written in the
doc/LICENSE file. This man page was written by Jason Edgecombe for
OpenAFS.


NEWS
OpenAFS News -- history of user-visible changes.

* Changes incorporated in OpenAFS 1.4.0

** Addition of pthreading support in volserver.

** Threading optimization of RX libraries.

** Windows 95/98 client desupported.

** Digital UNIX/Tru64 desupported.

** Windows XP/2003 client includes persistent (across boots) cache support.

** Windows XP/2003 now supports a cache of up to 1.2gb.

** Windows XP/2003 has various improvements to the included GUI and command 
   line tools.

** Windows XP/2003 now supports AFS UNC paths.

** MacOS X (10.3) and Linux use system inodes instead of a private inode pool.

** Files larger than 2gb are supported in Unix clients, and in servers if the 
   --enable-largefile-fileserver switch is enabled at configure time.

** Obsolete and insecure auxiliary network daemons (rcp, rsh, ftpd et al)
   have been removed.

** Fileserver optimizations have been made to improve timeout and callback 
   breaking behavior to inaccessible clients or clients whose address has
   changed.

** Kerberos 5 support is integrated.

** -nosettime is now the default for afsd.  Use "-settime" to get the
   old behavior.

** Mountpoint directory information is now only faked for cross-cell
   mountpoints when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd, but might not be stable
   on all platforms.

* Changes incorporated in OpenAFS 1.2.9

** The kaserver now defaults to not allowing interrealm authentication,
   due to security vulnerabilities in the krb4 protocol.  The new
   "-crossrealm" flag to the kaserver is provided to reenable interrealm
   authentication if desired.

** RedHat Linux 9.0 is now supported.

** Solaris 9 12/02 is now supported.  Solaris 7 and 8 x86 should now
   work again.

** On Linux machines using 2.2 series kernels, 2.2.19 or higher is now
   required.

** An OpenAFS 1.2.9 afsd will not work with kernel modules built from
   an earlier OpenAFS release.  In general, using a mismatched afsd and
   kernel modules set is unsupported; it is not recommended that you use
   such a configuration on a regular basis.

* Changes incorporated in OpenAFS 1.2.8

** Mountpoint directory information is now only faked for cross-cell
   mountpoits when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** HPUX 11.0 is now supported.

** It is now possible for AFS to use Kerberos 5 directly, via rxkad 2b.
   See the OpenAFS 1.2.8 Release Notes for more information on using this
   capability.

** An NFS translator kernel module is now included and compiled by default
   for Solaris only.

* Changes incorporated in OpenAFS 1.2.7

** MacOS X 10.2 is now supported.  FreeBSD 4.3 and later support is included
   in this release, but is still under active development and should only
   be used by those doing active development on the OpenAFS FreeBSD client.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** The fileserver will now use Rx pings to determine if clients are reachable
   prior to allocating resources to them, to prevent asymmetric clients from
   consuming all fileserver resources.

* Changes incorporated in OpenAFS 1.2.6

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd.

** Solaris 9 FCS and Solaris 7 and 8 x86 are now supported.

* Changes incorporated in OpenAFS 1.2.5

** A remote denial of service attack in the AIX and IRIX clients has
   been fixed.  Users of those platforms are strongly encouraged to
   upgrade.

** Fixed race conditions in fileserver that could result in crash.

* Changes incorporated in OpenAFS 1.2.4

** Server logfiles now more consistant about format in which hosts are
   referred to.

** vfsck on Solaris will now allow force runs (using -y flag) even if old 
   inodes exist.

* Changes incorporated in OpenAFS 1.2.3

** Cell aliases for dynroot can be specified in the CellAlias file in
   /usr/vice/etc or /usr/local/etc/openafs, in format "realname alias",
   one per line.  They can also be managed at runtime with "fs newalias"
   and "fs listaliases".

* Changes incorporated in OpenAFS 1.2.2

** Solaris 9 and Linux PA-RISC are now supported

** fileserver will not erroneously delay legitimate errors for 3 seconds
   after 10 errors are returned (e.g. stat() on a directory you can't read)

** Rx MTU calculation now works for Irix, Solaris and Linux

** If afsd is started with the -dynroot flag, /afs will be locally
   generated from the CellServDB.  AFSDB cells will be mounted
   automatically upon access.

** The namei fileserver allows vice "partitions" to be directories instead
   of partitions and will attach and display accordingly. Creating the file
   "AlwaysAttach" in the /vicepX directory is used as the trigger to attach it.

** TSM support for butc no longer requires editing a Makefile, simply
   specify the --enable-tivoli-tsm configure option.

** Linux builds no longer require source changes every time the kernel

src/butc/read_tape.c
/*
 * Copyright 2000, International Business Machines Corporation and others.
 * All Rights Reserved.
 * 
 * This software has been released under the terms of the IBM Public
 * License.  For details, see the LICENSE file in the top-level source
 * directory or online at http://www.openafs.org/dl/license10.html
 */

#include <afsconfig.h>
#include <afs/param.h>


#include <afs/cmd.h>
#include <lock.h>
#include <afs/tcdata.h>

#include <sys/types.h>
#include <netinet/in.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>

#include <afs/usd.h>

usd_handle_t fd;
usd_handle_t ofd;
int ofdIsOpen = 0;
afs_int32 nrestore, nskip;
int ask, printlabels, printheaders, verbose;
char filename[100];
char *outfile;

#define TAPE_MAGIC  1100000009	/* Defined in file_tm.c */
#define BLOCK_MAGIC 1100000005
#define FILE_MAGIC  1000000007
#define FILE_BEGIN     0
#define FILE_END       1
#define FILE_EOD      -1

struct tapeLabel {		/* also in file_tm.c */
    afs_int32 magic;
    struct butm_tapeLabel label;
};
struct fileMark {		/* also in file_tm.c */
    afs_int32 magic;
    afs_uint32 nBytes;
};

/* Read a tape block of size 16K */
afs_int32
readblock(buffer)
     char *buffer;
{
    static rerror = 0;
    u_int nread, total = 0;
    int rc, fmcount = 0;

    while (total < BUTM_BLOCKSIZE) {
	rc = USD_READ(fd, buffer + total, BUTM_BLOCKSIZE - total, &nread);
	if (rc != 0) {
	    return (rc);
	} else if ((nread == 0) && (total == 0)) {
	    if (verbose)
		fprintf(stderr, "Hardware file mark\n");
	    if (++fmcount > 3) {
		if (verbose)
		    fprintf(stderr,
			    "Greater than 3 hardware file marks in a row - done\n");
		return -1;
	    }
	} else if (nread == 0) {
	    fprintf(stderr, "Reached unexpected end of dump file\n");
	    return -1;
	} else {
	    total += nread;
	}
    }
    return 0;
}

void
printLabel(tapeLabelPtr)
     struct tapeLabel *tapeLabelPtr;
{
    tapeLabelPtr->label.dumpid = ntohl(tapeLabelPtr->label.dumpid);
    tapeLabelPtr->label.creationTime =
	ntohl(tapeLabelPtr->label.creationTime);
    tapeLabelPtr->label.expirationDate =
	ntohl(tapeLabelPtr->label.expirationDate);
    tapeLabelPtr->label.structVersion =
	ntohl(tapeLabelPtr->label.structVersion);
    tapeLabelPtr->label.useCount = ntohl(tapeLabelPtr->label.useCount);
    tapeLabelPtr->label.size = ntohl(tapeLabelPtr->label.size);

    fprintf(stderr, "\nDUMP       %u\n", tapeLabelPtr->label.dumpid);
    if (printlabels) {
	fprintf(stderr, "   AFS Tape  Name   : %s\n",
		tapeLabelPtr->label.AFSName);
	fprintf(stderr, "   Permanent Name   : %s\n",
		tapeLabelPtr->label.pName);
	fprintf(stderr, "   Dump Id          : %u\n",
		tapeLabelPtr->label.dumpid);
	fprintf(stderr, "   Dump Id Time     : %.24s\n",
		ctime(&(tapeLabelPtr->label.dumpid)));
	fprintf(stderr, "   Date Created     : %.24s\n",
		ctime(&(tapeLabelPtr->label.creationTime)));
	fprintf(stderr, "   Date Expires     : %.24s\n",
		ctime(&(tapeLabelPtr->label.expirationDate)));
	fprintf(stderr, "   Version Number   : %d\n",
		tapeLabelPtr->label.structVersion);
	fprintf(stderr, "   Tape Use Count   : %d\n",
		tapeLabelPtr->label.useCount);
	fprintf(stderr, "   Tape Size        : %u\n",
		tapeLabelPtr->label.size);
	fprintf(stderr, "   Comment          : %s\n",
		tapeLabelPtr->label.comment);
	fprintf(stderr, "   Dump Path        : %s\n",
		tapeLabelPtr->label.dumpPath);
	fprintf(stderr, "   Cell Name        : %s\n",
		tapeLabelPtr->label.cell);
	fprintf(stderr, "   Creator Name     : %s\n",
		tapeLabelPtr->label.creator.name);
	fprintf(stderr, "   Creator Instance : %s\n",
		tapeLabelPtr->label.creator.instance);
	fprintf(stderr, "   Creator Cell     : %s\n",
		tapeLabelPtr->label.creator.cell);
    }
}

void
printHeader(headerPtr, isvolheader)
     struct volumeHeader *headerPtr;
     afs_int32 *isvolheader;
{
    static volcount = 0;

    *isvolheader = 0;
    headerPtr->volumeID = ntohl(headerPtr->volumeID);
    headerPtr->server = ntohl(headerPtr->server);
    headerPtr->part = ntohl(headerPtr->part);
    headerPtr->from = ntohl(headerPtr->from);
    headerPtr->frag = ntohl(headerPtr->frag);
    headerPtr->magic = ntohl(headerPtr->magic);
    headerPtr->contd = ntohl(headerPtr->contd);
    headerPtr->dumpID = ntohl(headerPtr->dumpID);
    headerPtr->level = ntohl(headerPtr->level);
    headerPtr->parentID = ntohl(headerPtr->parentID);
    headerPtr->endTime = ntohl(headerPtr->endTime);
    headerPtr->versionflags = ntohl(headerPtr->versionflags);
    headerPtr->cloneDate = ntohl(headerPtr->cloneDate);

    if (headerPtr->magic == TC_VOLBEGINMAGIC) {
	*isvolheader = 1;
	if (verbose)
	    fprintf(stderr, "Volume header\n");
	fprintf(stderr,
		"VOLUME %3d %s (%u) - %s dump from %.24s till %.24s\n",
		++volcount, headerPtr->volumeName, headerPtr->volumeID,
		(headerPtr->level ? "Incr" : "Full"),
		((headerPtr->from) ? (char *)ctime(&headerPtr->from) : "0"),
		ctime(&(headerPtr->cloneDate)));
	if (printheaders) {
	    fprintf(stderr, "   Volume Name    = %s\n",
		    headerPtr->volumeName);
	    fprintf(stderr, "   Volume ID      = %u\n", headerPtr->volumeID);
	    fprintf(stderr, "   Clone Date     = %.24s\n",
		    ctime(&headerPtr->cloneDate));
	    fprintf(stderr, "   Vol Fragment   = %d\n", headerPtr->frag);
	    fprintf(stderr, "   Vol Continued  = 0x%x\n", headerPtr->contd);
	    fprintf(stderr, "   DumpSet Name   = %s\n",
		    headerPtr->dumpSetName);
	    fprintf(stderr, "   Dump ID        = %u\n", headerPtr->dumpID);
	    fprintf(stderr, "   Dump Level     = %d\n", headerPtr->level);
	    fprintf(stderr, "   Dump Since     = %.24s\n",
		    ctime(&headerPtr->from));
	    fprintf(stderr, "   parent Dump ID = %u\n", headerPtr->parentID);
	}
    } else if (headerPtr->magic == TC_VOLENDMAGIC) {
	if (verbose)
	    fprintf(stderr, "Volume Trailer\n");
    } else {
	fprintf(stderr, "Unrecognized Volume Header/Trailer\n");
    }
}

int
openOutFile(headerPtr)
     struct volumeHeader *headerPtr;
{
    afs_int32 len;
    int ch;
    int rc;
    int oflag;
    int skip, first;
    char rest[100];
    afs_hyper_t size;

    /* If we were asked to skip this volume, then skip it */
    if (nskip) {
	nskip--;
	return 0;
    }
    /* Skip if we are not to restore any */
    if (!nrestore)
	return 0;

    /* Get the volume name and strip off the BK or RO extension */
    if (outfile) {
	strcpy(filename, outfile);
    } else {
	strcpy(filename, headerPtr->volumeName);
	len = strlen(filename);
	if ((len > 7) && (strcmp(".backup", filename + len - 7) == 0)) {
	    filename[len - 7] = 0;
	} else if ((len > 9)
		   && (strcmp(".readonly", filename + len - 9) == 0)) {
	    filename[len - 9] = 0;
	}
    }

    if (ask) {
	first = 1;
	skip = 0;
	printf("Press return to retreive volume %s (%u) to file %s; " "s"
	       " to skip\n", headerPtr->volumeName, headerPtr->volumeID,
	       filename);
	do {
	    ch = getchar();
	    if ((first == 1) && (ch == 's'))
		skip = 1;
	    if ((first == 1) && (ch == 'q'))
		exit(0);
	    first = 0;
	} while (ch != '\n');
	if (skip) {
	    printf("Will not restore volume %s\n", headerPtr->volumeName,
		   headerPtr->volumeID);
	    return 0;
	}
    } else {
	printf("Retreive volume %s (%u) to file %s\n", headerPtr->volumeName,
	       headerPtr->volumeID, filename);
    }

    /* Should I append the date onto the end of the name? */

    /* Open the file to write to */
    if (headerPtr->contd == TC_VOLCONTD) {
	/* Continuation of dump */
	oflag = USD_OPEN_RDWR;
    } else {
	/* An all new dump */
	oflag = USD_OPEN_RDWR | USD_OPEN_CREATE;
    }
    rc = usd_Open(filename, oflag, 0664, &ofd);
    if (rc != 0) {
	fprintf(stderr, "Unable to open file %s. Skipping. Code = %d\n",
		filename, rc);
	nrestore--;
	return 0;
    }
    if (headerPtr->contd != TC_VOLCONTD) {
	hzero(size);
	rc = USD_IOCTL(ofd, USD_IOCTL_SETSIZE, &size);
	if (rc != 0) {
	    fprintf(stderr, "Unable to open file %s. Skipping. Code = %d\n",
		    filename, rc);
	    USD_CLOSE(ofd);
	    nrestore--;
	    return 0;
	}
    }
    ofdIsOpen = 1;
    return 0;
}

int
writeLastBlocks(char *lastblock, char *lastblock2)
{
    int rc;
    struct volumeHeader vh;
    char trailer[12];
    struct blockMark *bmark, *bmark2;
    char *data, *data2;
    int count, count2;
    int tlen, skip, pos;

    if (!ofdIsOpen)
	return 0;

    bmark = (struct blockMark *)lastblock;
    data = &lastblock[sizeof(struct blockMark)];
    count = ntohl(bmark->count);
    if (lastblock2) {
	bmark2 = (struct blockMark *)lastblock2;
	data2 = &lastblock2[sizeof(struct blockMark)];
	count2 = ntohl(bmark2->count);
    } else {
	count2 = 0;
    }

    /*
     * Strip off all but the last twelve bytes of the volume trailer
     */
    skip = sizeof(struct volumeHeader) - 12;
    if (count >= skip) {
	count = count - skip;
    } else if (count + count2 >= skip) {
	count2 = count2 - (skip - count);
	count = 0;
    } else {
	fprintf(stderr, "Failed to strip off volume trailer (1).\n");
	return 0;
    }

    /* volume trailer is somewhere in the last 12 bytes of the tape file.
     * The volume trailer may span tape blocks. */
    if (count >= 12) {
	tlen = 0;
	memcpy(trailer, data + (count - 12), 12);
    } else {
	tlen = 12 - count;
	memcpy(trailer, data2 + (count2 - tlen), tlen);
	if (count != 0) {
	    memcpy(trailer + tlen, data, count);
	}
    }

    for (pos = 0; pos <= 2; pos++) {
	if (strncmp(&trailer[pos], "H++NAME#", 8) == 0) {
	    break;
	}
	if (tlen > 0) {
	    tlen--;
	}
    }

    if (pos == 3) {
	fprintf(stderr, "Failed to strip off volume trailer (2).\n");
    } else {
	if (count2 - tlen > 0) {
	    rc = writeData(data2, count2 - tlen);
	}
	if ((tlen == 0) && (count > 12 - pos)) {
	    rc = writeData(data, count - (12 - pos));
	}
    }
    return 0;
}

int
closeOutFile()
{
    if (!ofdIsOpen)
	return 0;

    USD_CLOSE(ofd);
    ofdIsOpen = 0;

    /* Decrement the number of volumes to restore */
    nrestore--;
    return 0;
}

int
writeData(data, size)
     char *data;
     afs_int32 size;
{
    int rc;
    u_int nwritten;

    if (!ofdIsOpen)
	return 0;
    rc = USD_WRITE(ofd, data, (u_int) size, &nwritten);
    if (rc != 0) {
	fprintf(stderr, "Unable to write volume data to file. Code = %d\n",
		rc);
    }
    return 0;
}

static int
WorkerBee(struct cmd_syndesc *as, void *arock)
{
    char *tapedev;
    struct tapeLabel *label;
    struct fileMark *fmark;
    afs_int32 fmtype;
    struct blockMark *bmark;
    afs_int32 isheader, isdatablock;
    char *data;
    char *tblock;
    afs_int32 code;
    struct volumeHeader *volheaderPtr;
    int eod = 1;
    int rc;
    char *nextblock;		/* We cycle through three tape blocks so we  */
    char *lastblock;		/* can trim off the volume trailer from the  */
    char *lastblock2;		/* end of each volume without having to back */
    char *tapeblock1;		/* up the output stream.                     */
    char *tapeblock2;
    char *tapeblock3;

    tapedev = as->parms[0].items->data;	/* -tape     */
    nrestore =
	(as->parms[1].items ? atol(as->parms[1].items->data) : 0x7fffffff);
    nskip = (as->parms[2].items ? atol(as->parms[2].items->data) : 0);
    if (as->parms[4].items)
	nskip = 0x7fffffff;	/* -scan     */
    outfile = (as->parms[3].items ? as->parms[3].items->data : 0);
    ask = (as->parms[5].items ? 0 : 1);	/* -noask    */
    printlabels = (as->parms[6].items ? 1 : 0);	/* -label    */
    printheaders = (as->parms[7].items ? 1 : 0);	/* -vheaders */
    verbose = (as->parms[8].items ? 1 : 0);	/* -verbose  */

    /* Open the tape device */
    rc = usd_Open(tapedev, USD_OPEN_RDONLY, 0, &fd);
    if (rc != 0) {
	printf("Failed to open tape device %s. Code = %d\n", tapedev, rc);
	exit(1);
    }

    /*
     * Initialize the tape block buffers
     */
    tapeblock1 = (char *)malloc(3 * 16384);
    if (tapeblock1 == NULL) {
	printf("Failed to allocate I/O buffers.\n");
	exit(1);
    }
    tapeblock2 = tapeblock1 + 16384;
    tapeblock3 = tapeblock2 + 16384;

    nextblock = tapeblock1;
    lastblock = NULL;
    lastblock2 = NULL;

    /* Read each tape block deciding what to do with it */
    do {			/* while ((nskip!=0) && (nrestore!=0)) */
	code = readblock(nextblock);
	if (code) {
	    if (!eod)
		fprintf(stderr, "Tape device read error: %d\n", code);
	    break;
	}
	isdatablock = 0;

	/* A data block can be either a volume header, volume trailer,
	 * or actual data from a dump.
	 */
	bmark = (struct blockMark *)nextblock;
	label = (struct tapeLabel *)nextblock;
	fmark = (struct fileMark *)nextblock;
	if (ntohl(bmark->magic) == BLOCK_MAGIC) {
	    if (verbose)
		printf("Data block\n");
	    isdatablock = 1;
	    isheader = 0;
	    data = &nextblock[sizeof(struct blockMark)];
	    if (strncmp(data, "H++NAME#", 8) == 0) {
		volheaderPtr = (struct volumeHeader *)data;
		printHeader(volheaderPtr, &isheader);
	    }
	    if (isheader) {
		code = openOutFile(volheaderPtr);
		nextblock = tapeblock1;
		lastblock = NULL;
		lastblock2 = NULL;
	    } else {
		if (lastblock2 != NULL) {
		    data = &lastblock2[sizeof(struct blockMark)];
		    bmark = (struct blockMark *)lastblock2;
		    code = writeData(data, ntohl(bmark->count));
		    tblock = lastblock2;
		} else if (lastblock != NULL) {
		    tblock = tapeblock2;
		} else {
		    tblock = tapeblock3;
		}
		lastblock2 = lastblock;
		lastblock = nextblock;
		nextblock = tblock;
	    }
	}

	/* Filemarks come in 3 forms: BEGIN, END, and EOD.
	 * There is no information stored in filemarks.
	 */
	else if (ntohl(fmark->magic) == FILE_MAGIC) {
	    fmtype = ntohl(fmark->nBytes);
	    if (fmtype == FILE_BEGIN) {
		if (verbose)
		    fprintf(stderr, "File mark volume begin\n");
	    } else if (fmtype == FILE_END) {
		if (verbose)
		    fprintf(stderr, "File mark volume end\n");
	    } else if (fmtype == FILE_EOD) {
		if (verbose)
		    fprintf(stderr, "File mark end-of-dump\n");
		eod = 1;
	    }
	}

	/* A dump label */
	else if (ntohl(label->magic) == TAPE_MAGIC) {
	    if (verbose)
		fprintf(stderr, "Dump label\n");
	    printLabel(label);
	    eod = 0;
	}

	else {
	    if (verbose) {
		fprintf(stderr, "Unrecognized tape block - end\n");
	    }
	}

	/* Anything other than a data block closes the out file.
	 * At this point nextblock contains the end of tape file mark,
	 * lastblock contains the last data block for the current volume,
	 * and lastblock2 contains the second to last block for the current
	 * volume. If the volume fits in a single block, lastblock2 will
	 * be NULL. Call writeLastBlocks to strip off the dump trailer before
	 * writing the last of the volume data to the dump file. The dump
	 * trailer may span block boundaries.
	 */
	if (!isdatablock && lastblock) {
	    writeLastBlocks(lastblock, lastblock2);
	    closeOutFile();
	    nextblock = tapeblock1;
	    lastblock = NULL;
	    lastblock2 = NULL;
	}
    } while ((nskip != 0) || (nrestore != 0));

    free(tapeblock1);

    return 0;
}

main(argc, argv)
     int argc;
     char **argv;
{
    struct cmd_syndesc *ts;
    struct cmd_item *ti;

    setlinebuf(stdout);

    ts = cmd_CreateSyntax(NULL, WorkerBee, NULL,
			  "Restore volumes from backup tape");
    cmd_AddParm(ts, "-tape", CMD_SINGLE, CMD_REQUIRED, "tape device");
    cmd_AddParm(ts, "-restore", CMD_SINGLE, CMD_OPTIONAL,
		"# volumes to restore");
    cmd_AddParm(ts, "-skip", CMD_SINGLE, CMD_OPTIONAL, "# volumes to skip");
    cmd_AddParm(ts, "-file", CMD_SINGLE, CMD_OPTIONAL, "filename");
    cmd_AddParm(ts, "-scan", CMD_FLAG, CMD_OPTIONAL, "Scan the tape");
    cmd_AddParm(ts, "-noask", CMD_FLAG, CMD_OPTIONAL,
		"Prompt for each volume");
    cmd_AddParm(ts, "-label", CMD_FLAG, CMD_OPTIONAL, "Display dump label");
    cmd_AddParm(ts, "-vheaders", CMD_FLAG, CMD_OPTIONAL,
		"Display volume headers");
    cmd_AddParm(ts, "-verbose", CMD_FLAG, CMD_OPTIONAL, "verbose");

    return cmd_Dispatch(argc, argv);
}
